### PR TITLE
[SPARK-45034][SQL] Support deterministic mode function

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR adds a new optional argument to the `mode` aggregate function to provide deterministic results. When multiple values have the same greatest frequency then the new boolean argument can be used to get the lowest or highest value instead of an arbitraty one.

### Why are the changes needed?
To make the function more user friendly.

### Does this PR introduce _any_ user-facing change?
Yes, it adds a new argument to the `mode` function.

### How was this patch tested?
Added new UTs.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 1]
```diff
@@ -18,15 +18,22 @@
...
+import org.apache.spark.sql.catalyst.util.TypeUtils.toSQLExpr
+import org.apache.spark.sql.errors.DataTypeErrors.toSQLType
+import org.apache.spark.sql.types.{AbstractDataType, AnyDataType, ArrayType, BooleanType, DataType}
 import org.apache.spark.util.collection.OpenHashMap
 
 // scalastyle:off line.size.limit
 @ExpressionDescription(
-  usage = "_FUNC_(col) - Returns the most frequent value for the values within `col`. NULL values are ignored. If all the values are NULL, or there are 0 rows, returns NULL.",
+  usage = """
+    _FUNC_(col[, deterministicResult]) - Returns the most frequent value for the values within `col`. NULL values are ignored. If all the values are NULL, or there are 0 rows, returns NULL.
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: Maybe, make the parameter name shorter? How about `static` or `deterministic`?

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out [hunk 2]
```diff
@@ -1121,3 +1121,108 @@ struct<d:int>
...
+SELECT mode(col, true) FROM VALUES (map(1, 'a')) AS tab(col)
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.SparkIllegalArgumentException
+{
+  "errorClass" : "_LEGACY_ERROR_TEMP_2005",
+  "messageParameters" : {
+    "dataType" : "PhysicalMapType"
+  }
```
Comment (srielau) [inline sql/core/src/test/resources/sql-tests/results/group-by.sql.out:1227]: Would it be possible to fix that legacy error while we're in the neighborhood?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 1 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: +1 deterministic

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:35]: +1 for `deterministic`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 3]
```diff
@@ -81,7 +121,16 @@ case class Mode(
       return null
     }
 
-    buffer.maxBy(_._2)._1
+    (if (right.eval().asInstanceOf[Boolean]) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:124]: nit: we can follow `PercentileBase#percentages` and evaluate the foldable parameters in a lazy val.

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -1195,12 +1195,12 @@ def mode(col: "ColumnOrName") -> Column:
...
     ...     schema=("course", "year", "earnings"))
     >>> df.groupby("course").agg(mode("year")).show()
-    +------+----------+
-    |course|mode(year)|
-    +------+----------+
-    |  Java|      2012|
-    |dotNET|      2012|
-    +------+----------+
+    +------+-----------------+
+    |course|mode(year, false)|
```
Comment (MaxGekk) [inline python/pyspark/sql/functions.py:1199]: Shell we support new parameter in Python API too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 5]
```diff
@@ -35,24 +42,57 @@ import org.apache.spark.util.collection.OpenHashMap
...
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = super.checkInputDataTypes()
+    if (defaultCheck.isFailure) {
+      return defaultCheck
+    }
+    if (!deterministicResult.foldable) {
+      DataTypeMismatch(
+        errorSubClass = "NON_FOLDABLE_INPUT",
+        messageParameters = Map(
+          "inputName" -> "deterministicResult",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:87]: It is an id, so, should be quoted by `toSQLId`

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala]: This PR fixes other places: https://github.com/apache/spark/pull/42905

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:1227]: We have the ticket SPARK-42841 for this already. Could be replaced separately.

**Review by MaxGekk — APPROVED** [review]
LGTM except of minor comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 1 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala]: Ok, changed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 3 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala]: Sure. Fixed in latest.

File: python/pyspark/sql/functions.py (hunk 4 above)
Comment (peter-toth) [reply python/pyspark/sql/functions.py:774]: Thanks! This is actually a good question and I think we shall.
I've added the new param to both Scala and Python APIs in both SQL and Connect in the latest.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 5 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala]: Ok, fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 6]
```diff
@@ -35,24 +42,60 @@ import org.apache.spark.util.collection.OpenHashMap
...
+      DataTypeMismatch(
+        errorSubClass = "NON_FOLDABLE_INPUT",
+        messageParameters = Map(
+          "inputName" -> toSQLId("deterministic"),
+          "inputType" -> toSQLType(deterministicExpr.dataType),
+          "inputExpr" -> toSQLExpr(deterministicExpr)
+        )
+      )
+    } else {
+      TypeCheckSuccess
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:96]: shall we fail for null deterministic flag?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala (hunk 6 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:100]: ok, fixed in https://github.com/apache/spark/pull/42755/commits/948b8d7d8e478e84bfc2fe02cba9a73c8e0fed7e

**Comment by peter-toth** [comment]
Hmm, the failure seems unrelated but persistent...
```
[info] - client INVALID_CURSOR.DISCONNECTED error is retried when other RPC preempts this one *** FAILED *** (385 milliseconds)
[info]   org.apache.spark.SparkException: io.grpc.StatusRuntimeException: INTERNAL: [INVALID_CURSOR.POSITION_NOT_AVAILABLE] The cursor is invalid. The cursor position id c156089c-f861-467e-9031-31ee6f26d2b4 is no longer available at index 2.
[info]   at org.apache.spark.sql.connect.client.GrpcExceptionConverter$.toThrowable(GrpcExceptionConverter.scala:113)
```

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @peter-toth and @cloud-fan @srielau for review.

**Comment by peter-toth** [comment]
Thanks all for the review!

**Pull request closed without merging**